### ISSUE TITLE
[API Proposal]: Chunk method

### ISSUE THREAD

**Issue by Mr0N** [dotnet/runtime#101194]
### Background and motivation

I suggest adding a method of this kind to LINQ, in addition to what is currently available there.

### API Proposal

```csharp
using System.IO;

var en = Enumerable.Range(1, 88);
foreach (var item in Chunk(en, 8))
{
    Console.WriteLine("Items:------------");
    foreach (var element in item)
    {
        Console.WriteLine(element);
    }
}
Console.ReadKey();
IEnumerable<IEnumerable<T>> Chunk<T>(IEnumerable<T> en, int size)
{
    using var enumerator = en.GetEnumerator();
    bool check = true;
    do
    {
        var result = TryGetElements(enumerator);
        yield return result;
    }
    while (check);


    IEnumerable<T> TryGetElements(IEnumerator<T> enumerator)
    {
        for (int i = 0; i < size; i++)
        {
            if (!enumerator.MoveNext())
            {
                check = false;
                yield break;
            }
            yield return enumerator.Current;
        }
    }
}

```

### API Usage

```csharp

```

**Comment by stephentoub** [comment]
I don't understand the proposal. This:
```
public static IEnumerable<TSource[]> Chunk<TSource>(this IEnumerable<TSource> source, int size)
```
already exists. What are you proposing that's different, and why?

**Comment by Mr0N** [comment]
> I don't understand the proposal. This:
> 
> ```
> public static IEnumerable<TSource[]> Chunk<TSource>(this IEnumerable<TSource> source, int size)
> ```
> 
> already exists. What are you proposing that's different, and why?

I suggest adding an implementation from IEnumerable<IEnumerable<T>> to IEnumerable<T[]>, so that the Chunk method can be used without materializing the information into arrays first. Instead, these arrays could be yielded directly, allowing accessing the information directly from the IEnumerable being split. This means there's no need to write the information somewhere else before yielding it, thus saving memory.

I also suggest adding the following signature to the existing one:


public static IEnumerable<IEnumerable<T>> Chunk<TSource>(this IEnumerable<TSource> source, int size)

**Comment by stephentoub** [comment]
We can't overload by return type. Also, this signature was explicitly discussed and decided against when the existing Chunk was added.

**Comment by Mr0N** [comment]
Then the second method with another name, or for example reloading in the signature

public static IEnumerable<IEnumerable<T>> Chunk<TSource>(this IEnumerable<TSource> source, int size, book check)

**Comment by colejohnson66** [comment]
Why do you *need* `IEnumerable<IEnumerable<T>>`? This sounds like misguided optimization. Allocating a whole new enumerator each chunk would take longer and involve more allocations than a simple array.

**Comment by Mr0N** [comment]
> Why do you _need_ `IEnumerable<IEnumerable<T>>`? This sounds like misguided optimization. Allocating a whole new enumerator each chunk would take longer and involve more allocations than a simple array.

The semantics of IEnumerable lie in the fact that it can, for example, be 15 gigabytes, while the RAM is, for instance, 16, which can be a problem.

**Comment by colejohnson66** [comment]
Sure. But is that your actual use case? You probably shouldn't be using `IEnumerable<T>` for multiple gigabyte chunks. If it is indeed the best solution for your use case, it's a niche one; A custom extension method for your project is most likely the best solution.

```cs
public static class IEnumerableExtensions
{
    public static IEnumerable<IEnumerable<T>> ChunkEnumerable(this IEnumerable<T> source, int chunkSize);
}
```

**Comment by eiriktsarpalis** [comment]
I'll add that the proposed implementation has multiple `IEnumerable` instances sharing a single `IEnumerator` which is a potential minefield depending on how and in what order these enumerables are evaluated. I don't think this meets the bar for inclusion in LINQ.

**Comment by Mr0N** [comment]
> Member



> how and in what order these enumerables are evaluated.

Well, yeah, it's kind of an abstraction that allows both not to split and, at the same time, to split it.

**Comment by Mr0N** [comment]
Well, okay, you can create a lot of IEnumerable without actually reading them, well, that's probably more of a programmer's problem...

**Comment by eiriktsarpalis** [comment]
> Well, okay, you can create a lot of IEnumerable without actually reading them, well, that's probably more of a programmer's problem...

In the case of Linq we need to guarantee that each IEnumerable we return can be evaluated independently. This is why we never added `Unzip` methods for example.

**Comment by Mr0N** [comment]
> > Well, okay, you can create a lot of IEnumerable without actually reading them, well, that's probably more of a programmer's problem...
> 
> In the case of Linq we need to guarantee that each IEnumerable we return can be evaluated independently. This is why we never added `Unzip` methods for example.

So the Select method is not independent either, you can delete data from the collection and Select won't return anything.

**Comment by eiriktsarpalis** [comment]
> So the Select method is not independent either, you can delete data from the collection and Select won't return anything.

`Select` only returns one enumerable, the proposed method returns an enumerable of enumerables.

**Comment by colejohnson66** [comment]
Put it this way: there’s nothing stopping someone from *not* enumerating one of the inner ones. This would cause it to not chunk properly. The second inner enumerable would use data from that first one. The only way to properly chunk is to ensure that the outer enumerable is stepped *exactly* ‘count’ times. But you can’t do that unless you eagerly evaluate like the existing `Chunk` method does.

```
int[] source = [1, 2, 3, 4, 5];
IEnumerator<IEnumerable<int>> e = source.SpecialChunk(2);
e.MoveNext();
e.MoveNext();
int[] breakStuff = e.Current.ToArray();
Assert.IsTrue(breakStuff.SequenceEquals([1, 2]);
```

This breaks an invariant that enumerables should obey: not evaluating one “current” should not affect other values of “current”.

**Comment by colejohnson66** [comment]
> Well, this could happen in a multithreaded environment; this method isn't designed for multithreading.

I just showed how this can happen in a single-threaded environment.

> Well, and nobody guarantees the sequence.

What is the point of a chunking algorithm that can't chunk properly?

**Comment by Mr0N** [comment]
> What is the point of a chunking algorithm that can't chunk properly?

Because it doesn't violate the IEnumerable working principle, namely the MoveNext() method and the Current property; it takes one element from IEnumerable and doesn't store it in memory. Well, and it creates an abstraction that is pleasant to work with.

**Comment by CyrusNajmabadi** [comment]
I would completely assume I could concurrently read from all those inner enumerables.  That's literally how we use Chunk today.   To break up the input stream into chunks that we can then process in parallel. 

Note that my scenario involves no writing, just reading. If your approach doesn't let me read from arbitrary enumerables received from a make that is not being mutated, then it is broken as per the normal expectations around IEnumerable that have been around forever.

**Comment by Mr0N** [comment]
Perhaps this approach might not be very suitable for the Chunk method, considering the uncertainty regarding the number of IEnumerable that would be created and similar issues. Instead, such an approach would be better suited for a method like Split with a signature such as:


`public IEnumerable<IEnumerable<T>> Split(this IEnumerable<T> en, int countSegments);`
This method would take an enumeration as input and split it into a specified number of segments.

**Comment by Mr0N** [comment]
For example, if arrays are needed, you can always do:

`var res = Split(Enumerable.Range(1, 100), 10).Select(a => a.ToArray());

`

**Comment by colejohnson66** [comment]
Splitting is the same as chunking, just with a different name. If you really need enumerables, you can already do the following:

```cs
myEnumerable.Chunk(5).Select(e = e.AsEnumerable())
```

That gives the same result as your desired method, but is thread safe.

**Comment by Mr0N** [comment]
> Splitting is the same as chunking, just with a different name. If you really need enumerables, you can already do the following:
> 
> ```cs
> myEnumerable.Chunk(5).Select(e = e.AsEnumerable())
> ```
> 
> That gives the same result as your desired method, but is thread safe.

Well, the semantics are completely different, there are no data there, but there already are.
IEnumerable can be of such a type, for example:


public IEnumerable<string> GetDates(){
    yield return DateTime.Now.ToString();
    yield return DateTime.Now.ToString();
    yield return DateTime.Now.ToString();
    yield return DateTime.Now.ToString();
    yield return DateTime.Now.ToString();
    yield return DateTime.Now.ToString();

}

**Comment by Clockwork-Muse** [comment]
> If you really need enumerables, you can already do the following

You can already use arrays as `IEnumerable`s, so this doesn't really gain you anything.  The original request was about making the individual chunks lazy, which this doesn't provide.

That said...

The concerns about potential memory savings of providing a lazy chunk are likely overblown.  Consider that only a single chunk is created at a time - it doesn't realize the entire source collection.  I also suspect that an eager chunk is more useful to begin with - it enables passing it off for multithreaded processing.  If you make it lazy, all you're really doing at that point is provided an extra thing you have to call `MoveNext()` on... every once in a while.
